src/twisted/trial/_asynctest.py

```diff
@@ -131,43 +131,42 @@ def onTimeout(d):
     def __call__(self, *args, **kwargs):
         return self.run(*args, **kwargs)
 
-    def deferSetUp(self, ignored, result):
-        d = self._run("setUp", result)
-        d.addCallbacks(
-            self.deferTestMethod,
-            self._ebDeferSetUp,
-            callbackArgs=(result,),
-            errbackArgs=(result,),
-        )
-        return d
+    async def deferSetUp(self, result):
+        self._runNeedsTearDown = False
+        try:
+            try:
+                await self.deferSetUpNoCleanup(result)
+            finally:
+                await self.deferRunCleanups(None, result)
+        finally:
+            if self._runNeedsTearDown:
+                await self.deferTearDown(None, result)
 
-    def _ebDeferSetUp(self, failure, result):
-        if failure.check(SkipTest):
-            result.addSkip(self, self._getSkipReason(self.setUp, failure.value))
-        else:
-            result.addError(self, failure)
-            if failure.check(KeyboardInterrupt):
-                result.stop()
-        return self.deferRunCleanups(None, result)
-
-    def deferTestMethod(self, ignored, result):
```

Comment:
I assumed that `deferSetUp`, `deferTestMethod` and `deferTearDown` are effectively private anyway. There is no case where they would be useful to override. Also, their API is not "public ready" (i.e. the `ignored` parameter) - it seems as if they were some kind of internal thing that simply didn't get the `_` prefix.

Additionally, I don't think we can add a deprecation warning to these functions easily. We need to detect not when they are called, but when they are overridden. This means some kind of check whether `self.deferTestMethod == TestCase.deferTestMethod` either just before setUpClass or setUp, but even that wouldn't work if users override the method dynamically.

Furthermore, I don't think this PR can go in any form while preserving `deferSetUp`, `deferTestMethod` and `deferTearDown`. If we can't remove them without deprecation then the PR must wait until they can be removed.

Finally, I did search for these methods in GitHub and there's no public code that overrides them.

Given the above - small risk of breaking code for someone and relatively high costs for going through the deprecation dance I would suggest to have an exception for this change and allow to remove the functions without deprecation period.

I will need to update the release notes accordingly.

Please let me know what you think.

Reply:
Thanks for the info.
This is something that I have missed during the initial review.

-------


> There is no case where they would be useful to override.

This might be true.
But I still think that we should try to follow the "best practices" and handle this as removing public API.

We need to add a `12375.removal` newsfragment and inform Twisted trial users that this change was made.

And if we are doing the removal, I prefer to have the new names as private names... to reduce the need for any future communication.

And it will also need an email to the mailing list to inform about this change and ask for an exception.
I can help with mailing list communication

--------

As there are not many developers dedicating time to maintain `trial` I am +1 for making an exception here and removing these public methods without a deprecation warning.

SO I am +1 for approving the exception.

Reply:
> Given the above - small risk of breaking code for someone and relatively high costs for going through the deprecation dance I would suggest to have an exception for this change and allow to remove the functions without deprecation period.

I'm inclined to agree with your rationale for the exception, but please [follow the already-documented process for exceptions](https://docs.twisted.org/en/stable/development/compatibility-policy.html#procedure-for-exceptions-to-this-policy) :)

Reply:
Sure, I wasn't aware that I'm removing public API, as soon as Adi raised this the PR became totally wrong.

Reply:
I can help with sending the email to announce the process exceptions.

@p12tic as long as you think want to land this PR, I think that we can do it... it's just that the PR need a `.removal` newsfragment not a `.misc`

Reply:
@adiroiban I opened a separate PR for the removal https://github.com/twisted/twisted/pull/12389

Reply:
Addressed as part of https://github.com/twisted/twisted/pull/12389.